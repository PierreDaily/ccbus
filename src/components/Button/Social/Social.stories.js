import Social from "./index.vue";

// More on default export: https://storybook.js.org/docs/vue/writing-stories/introduction#default-export
export default {
  title: "Social",
  component: Social,
  // More on argTypes: https://storybook.js.org/docs/vue/api/argtypes
  argTypes: {
    app: {
      control: { type: "select" },
      options: ["twitter", "whatsapp", "facebook", "messenger"],
    },
  },
};

// More on component templates: https://storybook.js.org/docs/vue/writing-stories/introduction#using-args
const Template = (args) => ({
  // Components used in your story `template` are defined in the `components` object
  components: { Social },
  // The story's `args` need to be mapped into the template through the `setup()` method
  setup() {
    return { args };
  },
  // And then the `args` are bound to your component with `v-bind="args"`
  template: '<social v-bind="args" />',
});

export const Twitter = Template.bind({});
export const Facebook = Template.bind({});
export const Whatsapp = Template.bind({});
export const Messenger = Template.bind({});
// More on args: https://storybook.js.org/docs/vue/writing-stories/args

Twitter.args = {
  app: "twitter",
  url: "ccbus.pandy.community",
};

Facebook.args = {
  app: "facebook",
  url: "ccbus.pandy.community",
};

Whatsapp.args = {
  app: "whatsapp",
  url: "ccbus.pandy.community",
};

Messenger.args = {
  app: "messenger",
  url: "ccbus.pandy.community",
};
