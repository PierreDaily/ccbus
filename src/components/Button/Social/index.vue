<template>
    <a class="flex flex-col justify-start items-center w-18" :href="shareUrl" target="_blank" rel="noopener">
        <img class="w-16 h-16" :src="icon" />
        <span class="font-sans text-white text-base capitalize">{{ app }}</span>
    </a>
</template>

<script lang="ts">
import { defineComponent, computed } from 'vue';
import facebookIcon from "@/assets/icons/social/facebook.svg";
import messengerIcon from "@/assets/icons/social/messenger.svg";
import twitterIcon from "@/assets/icons/social/twitter.svg";
import whatsappIcon from "@/assets/icons/social/whatsapp.svg";

const FACEBOOK = 'facebook';
const MESSENGER = 'messenger';
const TWITTER = 'twitter';
const WHATSAPP = 'whatsapp';


export default defineComponent({
    name: 'Social',
    props: {
        app: {
            type: String,
            required: true,
            validator(value: string) {
                return [FACEBOOK, MESSENGER, TWITTER, WHATSAPP].includes(value)
            }
        },
        url: {
            type: String,
            required: true
        }
    },
    setup(props) {
        const icon = computed(() => {
            switch (props.app) {
                case FACEBOOK:
                    return facebookIcon

                case MESSENGER:
                    return messengerIcon

                case TWITTER:
                    return twitterIcon

                case WHATSAPP:
                    return whatsappIcon
            }
        })

        const shareUrl = computed(() => {
            switch (props.app) {
                case FACEBOOK:
                    return `http://www.facebook.com/share.php?u=${encodeURIComponent(props.url)}`

                case MESSENGER:
                    return `fb-messenger://share/?link=${encodeURIComponent(props.url)}`

                case TWITTER:
                    return `https://twitter.com/intent/tweet?url=${encodeURIComponent(props.url)}`

                case WHATSAPP:
                    return `https://api.whatsapp.com/send/?text=${encodeURIComponent(props.url)}`
            }
        })
        return {
            icon,
            shareUrl
        }
    }
})
</script>