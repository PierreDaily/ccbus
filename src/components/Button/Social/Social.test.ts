import { mount } from "@vue/test-utils";
import Social from "./index.vue";

test("Mount component", async () => {
  // arrange
  expect(Social).toBeTruthy();

  const wrapper = mount(Social, {
    props: {
      app: "facebook",
      url: "my-website.com",
    },
  });

  // assert
  expect(wrapper.html()).toMatchSnapshot();
});

test("Link to Facebook", async () => {
  // arrange
  const URL = "http://www.facebook.com/share.php?u=my-website.com";
  expect(Social).toBeTruthy();

  const wrapper = mount(Social, {
    props: {
      app: "facebook",
      url: "my-website.com",
    },
  });

  // assert
  expect(wrapper.find("a").exists()).toBe(true);
  expect(wrapper.find("a").attributes("href")).toBe(URL);
});

test("Link to Whatsapp", async () => {
  // arrange
  const URL = "https://api.whatsapp.com/send/?text=my-website.com";
  expect(Social).toBeTruthy();

  const wrapper = mount(Social, {
    props: {
      app: "whatsapp",
      url: "my-website.com",
    },
  });

  // assert
  expect(wrapper.find("a").exists()).toBe(true);
  expect(wrapper.find("a").attributes("href")).toBe(URL);
});

test("Link to Messenger", async () => {
  // arrange
  const URL = "fb-messenger://share/?link=my-website.com";
  expect(Social).toBeTruthy();

  const wrapper = mount(Social, {
    props: {
      app: "messenger",
      url: "my-website.com",
    },
  });

  // assert
  expect(wrapper.find("a").exists()).toBe(true);
  expect(wrapper.find("a").attributes("href")).toBe(URL);
});

test("Link to Twitter", async () => {
  // arrange
  const URL = "https://twitter.com/intent/tweet?url=my-website.com";
  expect(Social).toBeTruthy();

  const wrapper = mount(Social, {
    props: {
      app: "twitter",
      url: "my-website.com",
    },
  });

  // assert
  expect(wrapper.find("a").exists()).toBe(true);
  expect(wrapper.find("a").attributes("href")).toBe(URL);
});
