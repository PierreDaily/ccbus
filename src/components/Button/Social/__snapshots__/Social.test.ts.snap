// Vitest Snapshot v1

exports[`Mount component 1`] = `"<a class=\\"flex flex-col justify-start items-center w-18\\" href=\\"http://www.facebook.com/share.php?u=my-website.com\\" target=\\"_blank\\" rel=\\"noopener\\"><img class=\\"w-16 h-16\\" src=\\"/src/assets/icons/social/facebook.svg\\" /><span class=\\"font-sans text-white text-base capitalize\\">facebook</span></a>"`;
